Add sorting function to files list

jeffsieu opened the pull request:
**Files and folders can now be sorted** by clicking on
the column headers or by using the right-click context
menu. The sorting order mimics Windows Explorer,
by grouping the folders and files separately first, then
applying the relevant sort on the specified column,
then finally sorting by name.

Also:
- Improves right-click context menu action grouping
- Removes `RowIndex` from `ListedItem.cs` as it is not a
  reliable way to access items (indices change)
  after sorting is implemented
- Refactors code to work around the absence of `RowIndex`
- Other minor code refactoring

## Screenshots
![image](https://user-images.githubusercontent.com/8487294/67512843-456c5680-f6cc-11e9-8461-f7de5d0fbf38.png)
![image](https://user-images.githubusercontent.com/8487294/67512863-4f8e5500-f6cc-11e9-890d-94bce7cd7e46.png)
![image](https://user-images.githubusercontent.com/8487294/67512885-5a48ea00-f6cc-11e9-82c7-dc1a7ec2a963.png)

## Related Issues

Closes #178 

Possibly related: #20

yaichenbaum commented:
Do these changes apply to the photo view as well?

jeffsieu commented:
No, only on the default list view.

yaichenbaum commented:
Ok, I tested this and it works as expected, were you planning on doing it for photo view as well?
Side note, in the future we plan on merging the views and changing it to have different layout options so work like this wont need to be done more then once.

jeffsieu commented:
How does it work on photo view? Is it just via the right click menu, since there are no column headers? I haven't thought about doing it for the photo view as I don't usually use the photo view in Windows Explorer haha. But I guess I could get around to doing it some time soon.

duke7553 commented:
@jeffsieu Great work!

duke7553 commented:
My only concern would be if this breaks any functions that perform an action on a specific DataGrid index. This will require a bit of testing, still great to have regardless.

duke7553 commented:
@jeffsieu Looks great. After testing, I found several issues which we will need to address before I merge this branch to master:
1.) DataGrid doesn't remember sort options after navigation, but still displays the indicator on the wrong column
2.) Sorting by "Date modified" still needs work to be more accurate
3.) PhotoAlbum needs sort as well.
4.) Add Sort dropdown to ribbon in Layout tab

jeffsieu commented:
@duke7553 I've added sort by to PhotoAlbum and fixed 1). All that's left is making the sort dropdown in the ribbon.

As to 3), which part of sorting by "Date modified" is inaccurate?

yaichenbaum commented:
@duke7553 Can you explain what you meant by "Date modified" being inaccurate?
